src/saltext/vmware/modules/license_mgr.py

```diff
@@ -0,0 +1,207 @@
+
+    license_key
+        License Key which specifies license to add to license manager
+
+    service_instance
+        Use this vCenter service connection instance instead of creating a new one [default None]
+
+    datacenter_name
+        Datacenter name to use for the operation [default None]
+
+    cluster_name
+        Name of the cluster to add license [default None]
+
+    esxi_hostname
+        Hostname of the ESXI Server to add license [default None]
+
+    CLI Example:
+
+    .. code-block: bash
+
+        salt '*' vmware_license_mgr.add license_key datacenter_name=dc1
+    """
+    log.debug("start vmware ext license_mgr add license_key")
+    ret = {}
+    op = {}
+    for key, value in kwargs.items():
+        op[key] = value
+
+    service_instance = op.pop("service_instance", None)
+    datacenter_name = op.pop("datacenter_name", None)
+    cluster_name = op.pop("cluster_name", None)
+    esxi_hostname = op.pop("esxi_hostname", None)
+
+    if service_instance is None:
+        service_instance = get_service_instance(opts=__opts__, pillar=__pillar__)
+
+    if not utils_license_mgr.is_vcenter(service_instance):
+        ret["message"] = "Failed, not connected to a vCenter"
+        ret["result"] = False
+        return ret
+
+    try:
+        if "test" in __opts__ and __opts__["test"]:
+            ret["licenses"] = license_key
+            ret["message"] = "Test dry-run, not really connected to a vCenter testing"
+            return ret
+
+        result = utils_license_mgr.add_license(
+            service_instance, license_key, datacenter_name, cluster_name, esxi_hostname
+        )
+        if result:
+            ret["licenses"] = license_key
+
+    except (
+        salt.exceptions.VMwareApiError,
+        salt.exceptions.VMwareObjectRetrievalError,
+        salt.exceptions.VMwareRuntimeError,
+        salt.exceptions.CommandExecutionError,
+    ) as exc:
+        log.exception(exc)
+        ret[
+            "message"
+        ] = f"Failed to add a license key '{license_key}' due to Exception '{str(exc)}'"
+        ret["result"] = False
+        return ret
+
+    if not result:
+        ret[
+            "message"
+        ] = f"Failed specified license key '{license_key}' was not added to License Manager"
+        ret["result"] = False
+
+    return ret
+
+
+def list_(service_instance=None):
+    """
+    Returns a list of licenses for the specified Service Instance
+
+    service_instance
+        Use this vCenter service connection instance instead of creating a new one [default None]
+
+    .. code-block:: bash
+
+        salt '*' vmware_license_mgr.list
+    """
+    log.debug("start vmware ext license_mgr list")
+    if service_instance is None:
+        service_instance = get_service_instance(opts=__opts__, pillar=__pillar__)
+
+    if "test" in __opts__ and __opts__["test"]:
+        ret["licenses"] = "DGMTT-FAKED-TESTS-LICEN-SE012"
```

Comment:
So I'd have something to validity compare and asset against in tests

Reply:
Would it be a better approach to mock this out in the tests? Either the utils function, or mock out the service instance to return a known value?

For general usage, I would think that list would just return the licenses, regardless of test mode, since they would be needed in the state.

Reply:
removing the test section. as per conversation.